File: js/src/backends/convex.ts
import type { Tenant } from "../context.js";
import type { TenantBackend } from "./base.js";

export interface ConvexConfig {
  /** Convex deployment URL */
  deploymentUrl: string;
  /** API token for authentication (optional) */
  apiToken?: string;
}

/**
 * Convex backend for tenant storage using Convex HTTP API.
 */
export class ConvexBackend implements TenantBackend {
  private deploymentUrl: string;
  private apiToken?: string;

  constructor(config: ConvexConfig) {
    this.deploymentUrl = config.deploymentUrl;
    this.apiToken = config.apiToken;
  }

  private async callQuery<T>(functionName: string, args: Record<string, unknown>): Promise<T | null> {
    const headers: Record<string, string> = { "Content-Type": "application/json" };
    if (this.apiToken) {
      headers["Authorization"] = "Bearer " + this.apiToken;
    }

    const res = await fetch(this.deploymentUrl + "/api/query", {
      method: "POST",
      headers,
      body: JSON.stringify({ path: functionName, args }),
    });

    if (!res.ok) return null;
    const data = (await res.json()) as { value?: T };
    return data.value ?? null;
  }

  async getTenantByDomain(domain: string): Promise<Tenant | null> {
    return this.callQuery<Tenant>("tenants:getByDomain", { domain });
  }

  async getTenantById(id: string): Promise<Tenant | null> {
    return this.callQuery<Tenant>("tenants:getById", { id });
  }

  async listTenants(activeOnly = true): Promise<Tenant[]> {
    const result = await this.callQuery<Tenant[]>("tenants:list", { activeOnly });
    return result ?? [];
  }
}

File: py/multitenant/backends/convex.py
"""Convex backend for tenant storage via HTTP API."""

from typing import Any, Optional
import json
import urllib.request
import urllib.error

from django.conf import settings

from .base import TenantBackend


class ConvexBackend(TenantBackend):
    """Convex backend for tenant storage using Convex HTTP API."""

    def __init__(self):
        self.deployment_url = getattr(settings, "CONVEX_DEPLOYMENT_URL", None)
        if not self.deployment_url:
            raise ValueError("CONVEX_DEPLOYMENT_URL setting is required")
        self.api_token = getattr(settings, "CONVEX_API_TOKEN", None)

    def _call_query(self, function_name: str, args: dict) -> Any:
        url = self.deployment_url + "/api/query"
        headers = {"Content-Type": "application/json"}
        if self.api_token:
            headers["Authorization"] = "Bearer " + self.api_token

        payload = json.dumps({"path": function_name, "args": args}).encode()
        req = urllib.request.Request(url, data=payload, headers=headers)

        try:
            with urllib.request.urlopen(req, timeout=10) as resp:
                data = json.loads(resp.read().decode())
                return data.get("value")
        except urllib.error.HTTPError:
            return None

    def get_tenant_by_domain(self, domain: str) -> Optional[Any]:
        return self._call_query("tenants:getByDomain", {"domain": domain})

    def get_tenant_by_id(self, tenant_id: Any) -> Optional[Any]:
        return self._call_query("tenants:getById", {"id": tenant_id})

    def list_tenants(self, active_only: bool = True) -> list[Any]:
        result = self._call_query("tenants:list", {"activeOnly": active_only})
        return result if isinstance(result, list) else []
